atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -249,6 +249,19 @@ public RaftContext(
     started = true;
 
     addCommitListener(new AwaitingReadyCommitListener());
+
+    if (!raftLog.isEmpty() && term == 0) {
+      // This will only happen when metastore is empty because the node has just restored from a
+      // backup. Backup only contains the logs. Other case, where this can happen is when the meta
+      // file was manually deleted to recover from an unexpected bug.
+      // In both cases, we should not restart the term from 0 because the assumption in raft is that
+      // the term always increase. After restore, it is safe to restart the term at the last log's
+      // term. During the first election, the term will be incremented by 1.
+      // In the second case, it is possible that the actual term is higher. But it is still safe to
+      // set it to last log's term because the actual term will be set when this node gets the first
+      // message from other healthy replicas.
+      setTerm(raftLog.getLastEntry().term());
+    }
```

Comment:
:question: Is there are reason to do this so "late" in the constructor? I think ideally we'd do this in line 214, right after setting the term from the metastore and opening the log.

Reply:
Hm, I guess this interacts with verification of the snapshot and we have to call  `StateUtil.verifySnapshotLogConsistent`  first which can either throw if validation fails or even reset the log.

Reply:
Slightly unrelated but: when we detect a gap in `StateUtil.verifySnapshotLogConsistent`  (so log index >= snapshot index + 1) _and_ the log is empty, we reset the log. How can we have a log index > 1 but an empty log at the same time? :thinking:

Reply:
> Hm, I guess this interacts with verification of the snapshot and we have to call StateUtil.verifySnapshotLogConsistent first which can either throw if validation fails or even reset the log.

Yes. This is the reason.

Reply:
> How can we have a log index > 1 but an empty log at the same time?

This can happen due to snapshot replication + resetting the log. Resetting the log will result in an empty log with index > 1. We reset the log before committing the snapshot. If the node restarted while committing the snapshot, but reset was successful the node will end up in this state.